nameresolution/sqlite/sqlite.go

```diff
@@ -0,0 +1,301 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sqlite
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+
+	internalsql "github.com/dapr/components-contrib/internal/component/sql"
+	"github.com/dapr/components-contrib/nameresolution"
+	"github.com/dapr/kit/logger"
+)
+
+// ErrNoHost is returned by ResolveID when no host can be found.
+var ErrNoHost = errors.New("no host found with the given ID")
+
+// Internally-used error to indicate the registration was lost
+var errRegistrationLost = errors.New("host registration lost")
+
+type resolver struct {
+	logger         logger.Logger
+	metadata       sqliteMetadata
+	db             *sql.DB
+	gc             internalsql.GarbageCollector
+	registrationID string
+	closed         atomic.Bool
+	closeCh        chan struct{}
+}
+
+// NewResolver creates a name resolver that is based on a SQLite DB.
+func NewResolver(logger logger.Logger) nameresolution.Resolver {
+	return &resolver{
+		logger:  logger,
+		closeCh: make(chan struct{}),
+	}
+}
+
+// Init initializes the name resolver.
+func (s *resolver) Init(ctx context.Context, md nameresolution.Metadata) error {
+	err := s.metadata.InitWithMetadata(md)
+	if err != nil {
+		return err
+	}
+
+	connString, err := s.metadata.GetConnectionString(s.logger)
+	if err != nil {
+		// Already logged
+		return err
+	}
+
+	// Show a warning if SQLite is configured with an in-memory DB
+	if s.metadata.SqliteAuthMetadata.IsMemory() {
+		s.logger.Warn("Configuring name resolution with an in-memory SQLite database. Service invocation across differet apps will not work.")
+	} else {
+		s.logger.Infof("Configuring SQLite name resolution with path %s", connString[len("file:"):strings.Index(connString, "?")])
+	}
+
+	s.db, err = sql.Open("sqlite", connString)
+	if err != nil {
+		return fmt.Errorf("failed to create connection: %w", err)
+	}
+
+	// Performs migrations
+	err = performMigrations(ctx, s.db, s.logger, migrationOptions{
+		HostsTableName:    s.metadata.TableName,
+		MetadataTableName: s.metadata.MetadataTableName,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to perform migrations: %w", err)
+	}
+
+	// Init the background GC
+	err = s.initGC()
+	if err != nil {
+		return err
+	}
+
+	// Register the host and update in background
+	err = s.registerHost(ctx)
+	if err != nil {
+		return err
+	}
+
+	go s.renewRegistration()
+
+	return nil
+}
+
+func (s *resolver) initGC() (err error) {
+	s.gc, err = internalsql.ScheduleGarbageCollector(internalsql.GCOptions{
+		Logger: s.logger,
+		UpdateLastCleanupQuery: func(arg any) (string, any) {
+			return fmt.Sprintf(`INSERT INTO %s (key, value)
+				VALUES ('nr-last-cleanup', CURRENT_TIMESTAMP)
+				ON CONFLICT (key)
+				DO UPDATE SET value = CURRENT_TIMESTAMP
+					WHERE (unixepoch(CURRENT_TIMESTAMP) - unixepoch(value)) * 1000 > ?;`,
```

Comment:
This means records have a TTL of 1000s in the db?

Reply:
Ohhh that's a copy-paste mistake. Good catch.

```suggestion
					WHERE unixepoch(CURRENT_TIMESTAMP) - unixepoch(value)  > ?;`,
```